### ISSUE TITLE
Access decayed learning rate when using a learning rate scheduler

### ISSUE THREAD

**Issue by MathiesW** [keras-team/keras#20069]
I use the Adam optimizer and want to employ CosineDecay. The initial learning rate for the CosineDecay is `initial_learning_rate=0.0`, alpha is set to ``alpha=0.01`` and the ``warmup_target=1e-3``. I now want to log the learning rate with a callback. I subclassed the CSVLogger and added the ``model.optimizer.learning_rate`` to the logs. In the history, it shows a constant learning rate of ``1e-5``, which does not change over the epochs.

I read that it was possible with TensorFlow to access the current learning rate with ``model.optimizer._decayed_lr()``. This method does not exist in Keras 3.0. How can I access the current learning rate?

**Comment by ghsanti** [comment]
did you try accessing the learning rate in the way it's done here, right?

https://keras.io/guides/writing_your_own_callbacks/#learning-rate-scheduling

the key step is:

```py
class CustomLearningRateScheduler(keras.callbacks.Callback):
    def on_epoch_begin(self, epoch, logs=None):
        if not hasattr(self.model.optimizer, "learning_rate"):
            raise ValueError('Optimizer must have a "learning_rate" attribute.')
        # Get the current learning rate from model's optimizer.
        lr = self.model.optimizer.learning_rate
```

**Comment by MathiesW** [comment]
Hi @ghsanti 
I attached my code below. I use the exact same method like in your example.
````
class MyCSVLogger(callbacks.CSVLogger):
    def on_epoch_end(self, epoch, logs=None):
        try:
            current_lr = self.model.optimizer.learning_rate
            logs["learning_rate"] = float(
            backend.convert_to_numpy(current_lr)
        )
        except AttributeError:
            pass
        
        return super().on_epoch_end(epoch, logs)
````
As I said I subclass the CSVLogger and simply add the learning rate to the logs. It works flawless as long as no LR scheduler is used. When using cosine decay, the learning rate it reads from `self.mode.optimizer.learning_rate` does not seem to resemble the real learning rate but some constant (for me, `1e-5`, even though I defined a learning rate of `1e-3`).

**Comment by ghsanti** [comment]
Could you add a demo to make help easier?

I tried [here in this gist](https://gist.github.com/ghsanti/b81173914aa4a1ecd67f080c351d7b11) and if I understand correctly you must be doing something similar, but there the learning rate changes accordingly.

**Comment by Siddharth-Latthe-07** [comment]
learning rate is managed through the tf.keras.optimizers.schedules module. To log the current learning rate, you can define a custom callback. 
sample code:-
```
import tensorflow as tf
import csv

class LearningRateLogger(tf.keras.callbacks.Callback):
    def on_epoch_end(self, epoch, logs=None):
        logs = logs or {}
        lr = self.model.optimizer.learning_rate
        if isinstance(lr, tf.keras.optimizers.schedules.LearningRateSchedule):
            lr = lr(self.model.optimizer.iterations)
        logs['learning_rate'] = float(tf.keras.backend.get_value(lr))
        with open('learning_rate_log.csv', mode='a') as file:
            writer = csv.writer(file)
            writer.writerow([epoch, logs['learning_rate']])
        print(f"Epoch {epoch + 1}: Learning rate is {logs['learning_rate']}")

# Instantiate the learning rate schedule
initial_learning_rate = 0.0
alpha = 0.01
warmup_target = 1e-3
cosine_decay = tf.keras.optimizers.schedules.CosineDecay(
    initial_learning_rate=warmup_target,
    decay_steps=1000,
    alpha=alpha
)

# Instantiate the optimizer with the learning rate schedule
optimizer = tf.keras.optimizers.Adam(learning_rate=cosine_decay)

# Compile the model
model.compile(optimizer=optimizer, loss='your_loss_function')

# Fit the model with the custom callback
history = model.fit(your_data, your_labels, epochs=10, callbacks=[LearningRateLogger()])
```
Hope this helps, Thanks

**Comment by MathiesW** [comment]
Thank you for your responses. It turned out that everything is working as intended; you can access the current learning rate via ``self.model.optimizer.learning_rate``. I have mistaken the ``steps`` parameter in the cosine decay for ``epochs``, consequently, the learning rate was already fully decayed when I logged it via my callback.
